cmd/thanos/tools_bucket.go

```diff
@@ -1420,3 +1455,90 @@ func registerBucketRetention(app extkingpin.AppClause, objStoreConfig *extflag.P
 		return nil
 	})
 }
+
+func registerBucketUploadBlocks(app extkingpin.AppClause, objStoreConfig *extflag.PathOrContent) {
+	cmd := app.Command("upload-snapshot", "Upload Snapshot push blocks from the provided path to the object storeage.")
+
+	tbc := &bucketUploadBlocksConfig{}
+	tbc.registerBucketUploadBlocksFlag(cmd)
+
+	cmd.Setup(func(g *run.Group, logger log.Logger, reg *prometheus.Registry, _ opentracing.Tracer, _ <-chan struct{}, _ bool) error {
+		ctx := context.Background()
+		httpConfContentYaml, err := tbc.prometheus.httpClient.Content()
+		if err != nil {
+			return errors.Wrap(err, "getting http client config")
+		}
+		httpClientConfig, err := httpconfig.NewClientConfigFromYAML(httpConfContentYaml)
+		if err != nil {
+			return errors.Wrap(err, "parsing http config YAML")
+		}
+
+		httpClient, err := httpconfig.NewHTTPClient(*httpClientConfig, "thanos-tool")
+		if err != nil {
+			return errors.Wrap(err, "Improper http client config")
+		}
+
+		m := &promMetadata{
+			promURL: tbc.prometheus.url,
+			client:  promclient.NewWithTracingClient(logger, httpClient, "thanos-tool"),
+		}
+
+		err = runutil.Retry(2*time.Second, ctx.Done(), func() error {
+			if err := m.UpdateLabels(ctx); err != nil {
+				level.Warn(logger).Log(
+					"msg", "failed to fetch initial external labels. Is Prometheus running? Retrying",
+					"err", err,
+				)
+				return err
+			}
+
+			level.Info(logger).Log(
+				"msg", "successfully loaded prometheus external labels",
+				"external_labels", m.Labels().String(),
+			)
+			return nil
+		})
+
+		if err != nil {
+			return errors.Wrap(err, "initial external labels query")
+		}
+
+		if len(m.Labels()) == 0 {
+			return errors.New("no external labels configured on Prometheus server, uniquely identifying external labels must be configured; see https://thanos.io/tip/thanos/storage.md#external-labels for details.")
+		}
+
+		confContentYaml, err := objStoreConfig.Content()
+		if err != nil {
+			return err
+		}
+
+		bkt, err := client.NewBucket(logger, confContentYaml, component.Upload.String())
+		if err != nil {
+			return err
+		}
+
+		bkt = objstoretracing.WrapWithTraces(objstore.WrapWithMetrics(bkt, extprom.WrapRegistererWithPrefix("thanos_", reg), bkt.Name()))
+
+		// Ensure we close up everything properly.
+		defer func() {
+			if err != nil {
+				runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
+			}
+		}()
+
+		if err := promclient.IsWALDirAccessible(tbc.path); err != nil {
+			level.Error(logger).Log("err", err)
+		}
+
+		defer runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
+
+		s := shipper.New(logger, reg, filepath.Join(tbc.path, "snapshots", tbc.snapshotName), bkt, m.Labels, metadata.BucketUploadSource,
```

Comment:
Trying to understand why we still hardcode the snapshot path.
Does it work with general any TSDB paths?